engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java

```diff
@@ -491,19 +491,30 @@ private SeldonMessage queryREST(
         ResponseEntity<String> httpResponse =
             restTemplate.postForEntity(uri, request, String.class);
         try {
+          String response = httpResponse.getBody();
+          logger.debug(response);
+          SeldonMessage.Builder builder = SeldonMessage.newBuilder();
+          JsonFormat.parser().ignoringUnknownFields().merge(response, builder);
+          SeldonMessage seldonMessage = builder.build();
           if (httpResponse.getStatusCode().is2xxSuccessful()) {
-            SeldonMessage.Builder builder = SeldonMessage.newBuilder();
-            String response = httpResponse.getBody();
-            logger.debug(response);
-            JsonFormat.parser().ignoringUnknownFields().merge(response, builder);
-            return builder.build();
+            return seldonMessage;
           } else {
             logger.error(
                 "Couldn't retrieve prediction from external prediction server -- bad http return code: "
                     + httpResponse.getStatusCode());
-            throw new APIException(
-                APIException.ApiExceptionType.ENGINE_MICROSERVICE_ERROR,
-                String.format("Bad return code %d", httpResponse.getStatusCode()));
+            Status seldonMessageStatus = seldonMessage.getStatus();
+            if (seldonMessageStatus == null) {
+                throw new APIException(
+                        APIException.ApiExceptionType.ENGINE_MICROSERVICE_ERROR,
+                        String.format("Bad return code %d", httpResponse.getStatusCodeValue()));
+            }
+            else
+            {
+                APIException.ApiExceptionType exceptionType =
+                        APIException.ApiExceptionType.CUSTOMIZED_EXCEPTION;
+                exceptionType.setMessage(seldonMessageStatus.getReason());
```

Comment:
Since it's changing the message of the global `APIException.ApiExceptionType.CUSTOMIZED_EXCEPTION`, is this thread-safe? 

What would happen with a different thread using `APIException.ApiExceptionType.CUSTOMIZED_EXCEPTION`? Which message would it get?

Reply:
Good point, I got rid of the customized enum, and use a constructor instead to avoid such a concern.

Reply:
I'm not too familiar with Java's `enum`, but is it possible to create your own at runtime? For example, keeping a similar structure as how it was before the last commit, but doing instead:

```java
APIException.ApiExceptionType exceptionType =
    APIException.ApiExceptionType(CUSTOM_EXCEPTION, seldonMessageStatus.getReason(), 400);
throw new APIException(exceptionType, seldonMessageStatus.getInfo());
```

That way we would reduce the changes to the `APIException` class. I'm a bit wary that adding the new fields which were previously part of `APIException.ApiExceptionType` may be duplicating that functionality.

Reply:
I doubt we can,  since enum is static, and its constructor has to be private.

Reply:
I agree. 

Adding an inherited `CustomException` from `APIException` could be a third option, but it doesn't add much value. It's something that can be done in a future if needed.

BTW I really like the new simplified interface of `APIException` (without the `getExceptionType()` calls)! It also makes it easier to extend it in a future.